### DISCUSSION TITLE
Setup an Underworld 3 conda envrionment

### DISCUSSION THREAD

**Discussion by julesghub** [underworldcode/underworld3-archive#169] [Show and tell]
Command line instruction to setup a development conda environment for compiling and running underworld3. 
The setup creates a conda env with jupyter-lab installed and visualisation enabled for x86 and arm achitectures. 

### Create a fresh conda environment and install dependencies. 

```bash
conda create -n uw3_env
conda activate uw3_env
conda install jgiordani::uw3_env
export PYTHONPATH=$CONDA_PREFIX/lib/ 
```

### Compile Underworld 3 and all viz dependencies :rocket: 

```bash
cd /path/to/uw3 ; pip install -e .
./test.sh # run short tests - not all pass at present
pip install pyvista trame trame-vtk trame-vuetify trame-plotly
jupyter-lab 
```

### Notes:
* Works in parallel on a laptop, HPC hasn't been tested yet.
* Visualisation setup isn't optimal, but it works :smile: 
* Option to use osmesa for headless rendering. Not tested.
```bash
pip uninstall vtk
pip install --extra-index-url https://wheels.vtk.org vtk-osmesa
```
* Petsc version is based on main after the 3.20.5 release, i.e. sha: 9e121fd0b884c2488089c9399ee22b80abac3649. This includes required code for boundary integration that didn't make it into the 3.20.5 release. See https://gitlab.com/petsc/petsc/-/commit/49104dcdf81492e26905f4a528ffd1f11a9d9046#note_1759512043 
* h5py is parallel enabled.

**Comment by lmoresi** [comment]
I'm stuck with this because:


> Could not solve for environment specs
> The following package could not be installed
> └─ uw3_env is not installable because it requires
>    ├─ metis >=5.1.1,<5.2.0a0 , which can be installed;
>    └─ mumps-mpi >=5.6.2,<5.6.3.0a0 , which requires
>       └─ metis >=5.1.0,<5.1.1.0a0 , which conflicts with any installable versions previously reported.
> 

I tried conda and mamba and both behave the same.
Fresh environment with python=3.10 and the same with python=3.11

L

Prof Louis Moresi.
+61 2 6125 5590 or +61 4 0333 1413

Editing online to clean this up

**Comment by lmoresi** [comment]
Hi @julesghub - I'm also having problems when I do this on binder. I have, however, got a good recipe for the binder build which I can merge into dev. (Changes are all in the .binder directory)

**Comment by julesghub** [comment]
Yeah I can look into this

**Comment by lmoresi** [comment]
The binder version is working (finally). 

It's a bit tricky to figure out which environment applies to which build phase but I did get it to work by building the uw3_env into the environment.yml file (early in the list) and ALSO (important) by specifying python 3.11 which is not the default in the current binder OS.

It's working for the dev branch - you can check out all the examples that need surface integrals and they are OK. The PyVista visualisation is also working but I do advise client-side rendering with binder, otherwise it blows out the available memory quite easily:


```python
    pl.show(jupyter_backend='client')
```

This is set as the default but this does not seem to stick. We might need to force feed this to PyVista / binder for any 3D models or high-res cases.


PS @julesghub
